cdap-app-fabric/src/main/java/co/cask/cdap/metadata/LineageAdmin.java

```diff
@@ -160,8 +200,112 @@ public Lineage computeLineage(final StreamId sourceStream, long startMillis, lon
+          ProgramId workflowProgramId = new ProgramId(workflowProgramRunId.getNamespace(),
+                                                        workflowProgramRunId.getApplication(),
+                                                        workflowProgramRunId.getType(),
+                                                        workflowProgramRunId.getProgram());
+          Relation workflowRelation;
+          NamespacedEntityId data = relation.getData();
+          if (data instanceof DatasetId) {
+            workflowRelation = new Relation((DatasetId) data, workflowProgramId,
+                                              relation.getAccess(), RunIds.fromString(workflowProgramRunId.getRun()));
+          } else {
+            workflowRelation = new Relation((StreamId) data, workflowProgramId,
+                                              relation.getAccess(), RunIds.fromString(workflowProgramRunId.getRun()));
+          }
+          relationsNew.put(entry.getKey(), workflowRelation);
+        }
+      }
+    }
+    return relationsNew;
+  }
+
+  private Set<String> getWorkflowIds (Multimap<RelationKey, Relation> relations,
+                                      Map<ProgramRunId, RunRecordMeta> runRecordMap) throws NotFoundException {
+
+    final Set<String> workflowIDs = new HashSet<>();
+    for (Relation relation : Iterables.concat(relations.values())) {
+      RunRecordMeta runRecord = runRecordMap.get(
+        new ProgramRunId(relation.getProgram().getNamespace(), relation.getProgram().getApplication(),
+                         relation.getProgram().getType(), relation.getProgram().getProgram(),
+                         relation.getRun().getId()));
+      if (runRecord != null && runRecord.getProperties().containsKey("workflowrunid")) {
+        String workflowRunId = runRecord.getProperties().get("workflowrunid");
+        workflowIDs.add(workflowRunId);
+      }
+    }
+
+    return workflowIDs;
+  }
+
+  private Multimap<RelationKey, Relation> doComputeRollupLineage(Multimap<RelationKey,
+    Relation> relations) throws NotFoundException {
+
+    // Make a set of all ProgramIDs in the relations
+    Set<ProgramRunId> programRunIdSet = new HashSet<>();
+    for (Relation relation : Iterables.concat(relations.values())) {
+      programRunIdSet.add(new ProgramRunId(relation.getProgram().getNamespace(), relation.getProgram().getApplication(),
+                                        relation.getProgram().getType(), relation.getProgram().getProgram(),
+                                        relation.getRun().getId()));
+    }
+
+    // Get RunRecordMeta for all these ProgramRunIDs
+    final Map<ProgramRunId, RunRecordMeta> runRecordMap = store.getRuns(programRunIdSet);
+
+    // Get workflow Run IDs for all the programs in the relations
+    final Set<String> workflowIDs = getWorkflowIds(relations, runRecordMap);
+
+    // Get Program IDs for workflow Run IDs
+    Map<ProgramRunId, RunRecordMeta> workflowRunRecordMap =
```

Comment:
I have filed JIRA CDAP-7571 to store workflow program name in run record so that this scan is not necessary. Please add a TODO over here.